# **Ejercicio 5: Qualitative_Bankruptcy Data Set:**
### https://archive.ics.uci.edu/ml/datasets/Qualitative_Bankruptcy

### **a. Describe el problema y el objetivo que se busca**

Se presenta un problema de clasificación con el uso de 6 parametros recobrados de un análisis de expertos para determinar el estrato donde clasificar un grupo particular de usuarios determinando si estan o no en bancarrota.

Los atributos son categóricos y tienen 3 posibles valores:
   P-Positivo
   A-Average (Promedio)
   N-Negativo
   
La clasificación final es binaria, con 2 posibles valores:
   B-Bankrupt
   NB-No Bankrupt

Entre los posibles usos de este dataset esta orientado principalmente a autorizaciones de herramientas financieras, como oportunidades de inversión, creditos, hipotecas, etc.


### **b. Justifica si el problema requiere técnicas de balanceo de clases**



In [1]:
import pandas as pd
import seaborn as sns

path_5 = 'C:\TMP\Qualitative_Bankruptcy_DS.csv'
headers = ["IndustrialRisk","ManagementRisk","FinancialFlexibility","Credibility","Competitiveness","OperatingRisk","Class"]

df_5= pd.read_csv(path_5, names=headers, sep= ',')
df_5.drop(index=df_5.index[0], 
        axis=0, 
        inplace=True)
df_5.head()

IndustrialRisk ManagementRisk FinancialFlexibility Credibility  \
1              P              P                    A           A   
2              N              N                    A           A   
3              A              A                    A           A   
4              P              P                    P           P   
5              N              N                    P           P   

  Competitiveness OperatingRisk Class  
1               A             P    NB  
2               A             N    NB  
3               A             A    NB  
4               P             P    NB  
5               P             N    NB

In [2]:
df_5.shape

(250, 7)

In [3]:
df_5.dtypes

IndustrialRisk          object
ManagementRisk          object
FinancialFlexibility    object
Credibility             object
Competitiveness         object
OperatingRisk           object
Class                   object
dtype: object

In [4]:
df_5.describe()

IndustrialRisk ManagementRisk FinancialFlexibility Credibility  \
count             250            250                  250         250   
unique              3              3                    3           3   
top                 N              N                    N           N   
freq               89            119                  119          94   

       Competitiveness OperatingRisk Class  
count              250           250   250  
unique               3             3     2  
top                  N             N    NB  
freq               103           114   143

In [5]:
df_5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 250 entries, 1 to 250
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   IndustrialRisk        250 non-null    object
 1   ManagementRisk        250 non-null    object
 2   FinancialFlexibility  250 non-null    object
 3   Credibility           250 non-null    object
 4   Competitiveness       250 non-null    object
 5   OperatingRisk         250 non-null    object
 6   Class                 250 non-null    object
dtypes: object(7)
memory usage: 15.6+ KB


Hacemos el tratamiento de los datos a valores numericos para el clasificador.
La etiqueta POSITIVO (P), PROMEDIO (A), NEGATIVO (N) cambian a [1,0,-1].

In [6]:
df_5Int = df_5

In [7]:
 df_5Int = df_5Int.replace(
     to_replace=['P','A','N'], 
     value=[1, 0,-1])

La etiqueta de las clases tambien se cambian
Bankruptcy (B), Non-Bankruptcy (NB) cambian a [1,0]

In [8]:
 df_5Int['Class'] = df_5Int['Class'].replace(
     to_replace=['NB','B'], 
     value=[0, 1])

In [9]:
df_5Int.head()

IndustrialRisk  ManagementRisk  FinancialFlexibility  Credibility  \
1               1               1                     0            0   
2              -1              -1                     0            0   
3               0               0                     0            0   
4               1               1                     1            1   
5              -1              -1                     1            1   

   Competitiveness  OperatingRisk  Class  
1                0              1      0  
2                0             -1      0  
3                0              0      0  
4                1              1      0  
5                1             -1      0

In [10]:
df_5Int.tail()

IndustrialRisk  ManagementRisk  FinancialFlexibility  Credibility  \
246              -1              -1                     0           -1   
247               1              -1                    -1           -1   
248               0              -1                    -1           -1   
249              -1              -1                    -1           -1   
250               1              -1                    -1           -1   

     Competitiveness  OperatingRisk  Class  
246               -1             -1      1  
247               -1             -1      1  
248               -1             -1      1  
249               -1             -1      1  
250                0              0      1

In [11]:
df_5Int.describe()

IndustrialRisk  ManagementRisk  FinancialFlexibility  Credibility  \
count      250.000000       250.00000            250.000000   250.000000   
mean        -0.036000        -0.22800             -0.248000    -0.060000   
std          0.823051         0.82141              0.803166     0.831363   
min         -1.000000        -1.00000             -1.000000    -1.000000   
25%         -1.000000        -1.00000             -1.000000    -1.000000   
50%          0.000000         0.00000              0.000000     0.000000   
75%          1.000000         0.00000              0.000000     1.000000   
max          1.000000         1.00000              1.000000     1.000000   

       Competitiveness  OperatingRisk       Class  
count       250.000000      250.00000  250.000000  
mean         -0.048000       -0.14000    0.428000  
std           0.881364        0.86915    0.495781  
min          -1.000000       -1.00000    0.000000  
25%          -1.000000       -1.00000    0.000000  
50%           0.000000        0.00000    0.000000  
75%           1.000000        1.00000    1.000000  
max           1.000000        1.00000    1.000000

In [12]:
df_5Int['Class'].value_counts()

0    143
1    107
Name: Class, dtype: int64

Dentro del análisis realizado encontramos un dataset sin problema de balanceo de clases dado que la clase positiva (en bancarrota) representa un 42.8% y la clase negativa (no en bancarrota) el 57.2%. Por ese motivo no es necesario realizar algun acondicionamiento del dataset para balancear clases.

## **Analiza las implicaciones que conllevan los FN y FP**

Para este dataset es importante discrimintar los falsos negativos (FN) ya que si esta información es la entrada para autorizar algun tipo de herramienta financiera la perdida seria cuantiosa, en manera similiar a los ejercicios propuestos en clase, como el data set de tarjeta de credito, para las cuestiones financieras tiene mayor impacto los FN.
Por otra parte, los FP, simplemente se perderían algunos clientes pero el riesgo de perdidas es menor.

## **Decide y justifica la métrica a utilizar**

Para el uso de este dataset, el costo de los FN es más crítico por lo que la métrica más adecuada para calificar el modelo debe considerar el FN. 

La métrica recomendada que se usa para evaluar las prediciones Negativas es Recall. 

Recall = ( TP / TP + FN )

También para tener un balance, la metrica que podría considerarse es la función F1, definida como:

F1 = 2 * ((recall * precision)/(recall + precision))